### ISSUE TITLE
A bug in anchor scales?

### ISSUE THREAD

**Issue by qinc** [STVIR/pysot#300]
Please correct me if I'm wrong for the following observations:

<strike>Based on @zyc-ai 's description in issue 129 https://github.com/STVIR/pysot/issues/129, as well as the paper, anchors should be generated based on the size of template bounding box, but</strike> in the code, wherever Anchors is generated it was always through code like this:
        self.anchors = Anchors(cfg.ANCHOR.STRIDE,
                               cfg.ANCHOR.RATIOS,
                               cfg.ANCHOR.SCALES)
which will give anchors with constant size (64x64, before applying ratios). For generating positive classified anchors, one anchor box need to have >0.6 IOU with the ground truth, that means if the ground truth bounding box is much bigger or smaller than 64x64, no anchor will be considered positive during training, and that object won't be detected during tracking at all.

A workaround is to defined multiple anchor scales, but the paper claims only one scale is needed, which is very confusing. This will only work if all the tracking objects are around the same 64x64 size in the search image.

Am I missing something? Please point out if there is code somewhere I've missed, thanks~!

_Originally posted by @qinc in https://github.com/STVIR/pysot/issues/129#issuecomment-590620000_

**Comment by zyc-ai** [comment]
Hi, I just reread your question. 

There is no relation between anchor and template.

**Comment by qinc** [comment]
Hi Zhiyuan,

I've reviewed my questions and yes there should be no relation between
anchor and template, I will edit that part later.

Thanks,
Chuan

**Comment by qinc** [comment]
But my main concern still holds: if there is no enough diversity in anchor scales, samples with very large or very small bounding box will be excluded during training. That means using more scales should improve the result of this work.

**Comment by zyc-ai** [comment]
> But my main concern still holds: if there is no enough diversity in anchor scales, samples with very large or very small bounding box will be excluded during training. That means using more scales should improve the result of this work.

anchors are generated based on the bounding box. and it might increase the performant, but it'll definitely increase the time of process very significantly too, which is unacceptable.

**Comment by qinc** [comment]
> anchors are generated based on the bounding box.

Anchors are NOT generated based on the bounding box. They were generated only once for all training samples. Only the positive/negative activation is based on each bounding box.

> and it might increase the performant, but it'll definitely increase the time of process very significantly too, which is unacceptable.

It may slightly increase training time but definitely not significant since there is always an upper-bound on for the number of activated anchors. And having more positive anchors will definitely help training, since for each sample negative anchors are always dominating. At inference time there is only marginal performance difference too.

But I have to admit in my experiment using just single scale anchor seems working for detecting objects with very large bounding box, even when that size should have been excluded during training. It seems the network has the capability to scale to those cases. Anyway it's just in my case so it's not anything conclusive, and I ended using multiple anchors scales for better supporting various object sizes, at least in theory.

**Comment by zyc-ai** [comment]
> Anchors are NOT generated based on the bounding box. They were generated only once for all training samples. Only the positive/negative activation is based on each bounding box.

Yes that's correct, anchors are generated on the feature maps. Sorry about the wrong reply. 

> And having more positive anchors will definitely help training
I highly doubt so. There is a reason why 
> negative anchors are always dominating

> I ended using multiple anchors scales for better supporting various object sizes, at least in theory.

It is actually a hyper parameter, and highly depends on your actual scenerio. Experiment is the only way to get truth.

**Comment by zyc-ai** [comment]
Since no further question is asked, I'm closing this issue for now. 
Please create a new issue should you have other questions.
